iocore/net/quic/QUICCongestionController.cc

```diff
@@ -36,7 +36,9 @@
         "window: %" PRIu32 " bytes: %" PRIu32 " ssthresh: %" PRIu32 " " fmt, \
         this->_info->cids().data(), this->_congestion_window, this->_bytes_in_flight, this->_ssthresh, ##__VA_ARGS__)
 
-QUICCongestionController::QUICCongestionController(QUICConnectionInfoProvider *info, const QUICCCConfig &cc_config) : _info(info)
+QUICCongestionController::QUICCongestionController(QUICRTTMeasure *rtt_measure, QUICConnectionInfoProvider *info,
```

Comment:
Can we pass `QUICRTTProvider` instead and make this parameter `const &`?

Reply:
I don't think pass QUICRTTProvider is good idea. because the congestion period needs to some of QUICRTTMeasure's member.

```
 ink_hrtime
 QUICRTTMeasure::congestion_period(uint32_t threshold) const
 {
   ink_hrtime pto = this->_smoothed_rtt + std::max(this->_rttvar * 4, this->_k_granularity);
   return pto * (1 << (threshold - 1));
 }

```

That's mean we need to expose the k_granularity api.


And I wants to make period calculate into QUICRTTMeasure so RTTMeasure could get any period like (congestion period, PTO, handshake timeout etc..)

```
  // period
   ink_hrtime handshake_retransmit_timeout() const;
   ink_hrtime current_pto_period() const;
   ink_hrtime congestion_period(uint32_t threshold) const;

   // get members
   ink_hrtime smoothed_rtt() const override;
   ink_hrtime rttvar() const override;
   ink_hrtime latest_rtt() const override;

   uint32_t pto_count() const;
   uint32_t crypto_count() const;

   void set_crypto_count(uint32_t count);
   void set_pto_count(uint32_t count);

   void update_rtt(ink_hrtime latest_rtt, ink_hrtime ack_delay);
   void reset();
```

Reply:
or make these function virtual .

Reply:
`QUICRTTProvider` is just an interface that makes writing mocks easy, and all logic will remain in `QUICRTTMeasure`. My suggestion is that declare all `const` functions (getters) in `QUICRTTProvider` as pure virtual functions, and implement them in `QUICRTTMeasure`.

Having xxx_period functions in `QUICRTTProvider` sounds odd because it's *RTT* provider, however you can say the same thing against `QUICRTTMeasure`. You can change the name if you want.

Do CC need to call any `RTTMeasure`'s non-const functions? I'm also ok with just making the parameter `const`.

Reply:
done